### PR TITLE
Niad-2913: Guidance around what happens with PS Adaptor in failure modes

### PR DESCRIPTION
**Pull request by ole4ryb**
## What

A guidance was added that highlights the specific focus on the dependent services and how the PS Adaptor operates when those services are unavailable

## Why

This guide outlines the expected behavior of the PS Adaptor when various dependent services are down. 
The PS Adaptor works in conjunction with several services including the GP2GP Translator, MHS Adaptor, Facade, and others. 
The guide explains the effects of outages and how the system behaves in each case that might be very helpful while investigating different issues related to the PS Adaptor.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Internal change (non-breaking change with no effect on the functionality affecting end users)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+## Service Failures and Expected Behavior
+
+This guide outlines the expected behavior of the PS Adaptor when various dependent services are down. 
+The PS Adaptor works in conjunction with several services including the GP2GP Translator, MHS Adaptor, Facade, and others.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:4]: This is confusing, the Translator and Facade are parts of the PS Adaptor?! This sentence suggests that they are separate. 😕

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+The guide explains the effects of outages and how the system behaves in each case.
+
+### System Overview
+The PS Adaptor initiates a conversation for transferring patient records, relying on several interconnected services:
+- GP2GP Translator: Sends patient records to the PS Adaptor via a queue.
+- Facade Application: Works with the PS Adaptor to manage transfer requests.
+- MHS Adaptor: Manages inbound and outbound queues, which are responsible for transferring patient records to the incumbent system.
+- Other Dependencies: Include PS Adaptor database, MHS inbound and outbound services, file storage, Redis cache, and MHS outbound SDS.
+
+### Service Outage Scenarios and Expected Behavior
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:14]: I find the word "expected" used here and below a bit confusing. What is the use of this word trying to achieve?

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+1. GP2GP Translator is Down
+   - Scenario: The GP2GP Translator is not operational and cannot send patient records.
+   - Expected Behavior: No messages are sent to the incumbent system while the translator is down. Once the GP2GP Translator recovers, 
+     the transfer process resumes, and messages are processed as normal.
+   - Recovery: Automatic upon Translator recovery; transfer resumes without manual intervention.
+2. Queue is Down
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:20]: 😕 Not sure if a "queue" can go down. Is this more the scenario that the [broker](https://en.wikipedia.org/wiki/Message_broker) goes down?

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+   - Scenario: The queue responsible for transferring data between the GP2GP Translator and PS Adaptor is down.
+   - Expected Behavior: The initial request is never sent to the incumbent system because the queue is unavailable.
+   - Recovery: After the queue is restored, the request can be processed normally, but any messages queued during downtime are not sent.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:23]: I'm very confused by this part. Can you help me understand what you mean by this?

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+3. PS Adaptor Database (DB) is Down
+   - Scenario: The PS Adaptor's database is not operational.
+   - Expected Behavior: The initial request to the incumbent system is not sent because the adaptor cannot access the necessary data.
+   - Recovery: After the database is restored, the system may resume operations, but requests during the outage are not processed.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:27]: Again I'm confused by the wording "initial request to the incumbent system is not sent"

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+4. Facade Application is Down
+   - Scenario: The facade application that works with the PS Adaptor is down.
+   - Expected Behavior: The initial request to the incumbent system is not sent. The communication between the PS Adaptor and the incumbent system is blocked.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:30]: Confused still as above.

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+   - Recovery: Once the facade application is operational, the system can process requests as normal.
+5. MHS Adaptor is Down
+   - Scenario: The MHS Adaptor (including its database) is not operational.
```
[the remark this answers is not in the document]
Comment (adrianclay) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:33]: Reading the sentence below, the scenario seems more to do with the MHS database specifically?! Which seems like a strange boundary to have.

From what I understand of the PS adaptor, and what I'm reading here, we could simplify this scenario to "MHS Outbound adaptor is down", and not be specific about what is causing it to have downtime.

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+   - Expected Behavior: The EHR request is not sent to the incumbent system while the MHS Adaptor is down. 
+     Once the MHS Adaptor's database recovers, the request is sent as normal.
+   - Recovery: Automatic upon recovery of the MHS Adaptor’s database.
+6. MHS Adaptor Inbound Queue is Down
+   - Scenario: The inbound queue managed by the MHS Adaptor is down.
+   - Expected Behavior: The EHR request is successfully sent to the incumbent system.
+   - Recovery: No action is required as the request is still processed.
+7. MHS Adaptor Outbound Queue is Down
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:41]: `Outbound queue?!`

I don't even know what the outbound queue is. Do you?

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+4. Facade Application is Down
+   - Scenario: The facade application that works with the PS Adaptor is down.
+   - Expected Behavior: The initial request to the incumbent system is not sent. The communication between the PS Adaptor and the incumbent system is blocked.
+   - Recovery: Once the facade application is operational, the system can process requests as normal.
+5. MHS Adaptor is Down
+   - Scenario: The MHS Adaptor (including its database) is not operational.
+   - Expected Behavior: The EHR request is not sent to the incumbent system while the MHS Adaptor is down. 
+     Once the MHS Adaptor's database recovers, the request is sent as normal.
+   - Recovery: Automatic upon recovery of the MHS Adaptor’s database.
+6. MHS Adaptor Inbound Queue is Down
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:37]: Similar to the point above, queues don't really go down but the broker can.

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+     However, any operations that rely on file storage will fail.
+   - Recovery: File storage recovery may be necessary for non-EHR-related operations, but the EHR transfer proceeds.
+9. Redis is Down
+   - Scenario: Redis, used for caching, is not operational.
+   - Expected Behavior: There are no negative effects on the transfer process. The translation of patient records continues as normal.
+   - Recovery: No specific recovery action is required for Redis as its absence does not hinder the transfer process.
+10. MHS Outbound SDS is Down
+    - Scenario: The MHS outbound SDS is down.
+    - Expected Behavior: The request is never sent to the incumbent system.
+    - Recovery: Once the MHS outbound SDS is operational again, the request will need to be manually retried or reprocessed.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:58]: Ooooh, probably need to be a bit clear about what is meant by this. What does "manually retried" mean?

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+8. File Storage is Down
+   - Scenario: The file storage system is not operational.
+   - Expected Behavior: Despite the outage, the EHR request is still sent to the incumbent system. 
+     However, any operations that rely on file storage will fail.
+   - Recovery: File storage recovery may be necessary for non-EHR-related operations, but the EHR transfer proceeds.
+9. Redis is Down
+   - Scenario: Redis, used for caching, is not operational.
+   - Expected Behavior: There are no negative effects on the transfer process. The translation of patient records continues as normal.
+   - Recovery: No specific recovery action is required for Redis as its absence does not hinder the transfer process.
+10. MHS Outbound SDS is Down
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:55]: This is an internal implementation detail of the MHS, could this be wrapped up within the "MHS Outbound" is down scenario?

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 1 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:4]: ok, I will correct it

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 3 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:20]: Yes, agreed as it is more likely scenario and more often it is a message broker itself that is down rather than a queue, however there are some scenarios when the queue itself can go down while the message broker could be still running and operational. For example, if the queue reaches its maximum capacity based on some config it becomes unreachable. There are also some other failures that can happen to the queue.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 2 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:14]: "Expected" meaning that we expect the components to react or behave in a particular manner once a specific situation occurs...

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+7. MHS Adaptor Outbound Queue is Down
+   - Scenario: The outbound queue managed by the MHS Adaptor is down.
+   - Expected Behavior: The request is not sent to the incumbent system while the outbound queue is down. 
+     Once the outbound service recovers, the request is processed and sent.
+   - Recovery: The system automatically sends the request when the outbound queue becomes operational again.
+8. File Storage is Down
+   - Scenario: The file storage system is not operational.
+   - Expected Behavior: Despite the outage, the EHR request is still sent to the incumbent system. 
+     However, any operations that rely on file storage will fail.
+   - Recovery: File storage recovery may be necessary for non-EHR-related operations, but the EHR transfer proceeds.
```
[the remark this answers is not in the document]
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:45]: No, we saying that as per the test that was done in NIAD-2369 the fact that the file system was down didn't affect the service however potentially it might have some effects on the transfer especially in those cases (which might not have been included in the test) that rely on the file storage.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 9 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: Corrected.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 8 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: Corrected.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 3 above)
Comment (adrianclay) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: Hmm, I'd recommend being specific about the particular scenario actually tested.

So if we've tested the scenario of "Broker dies", then let's document the behaviour then. If we've tested the scenario of "Broker alive, but not able to save the message", then let's document that behaviour.

For any other scenarios which aren't tested, let's not have them listed here.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 4 above)
Comment (adrianclay) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:23]: Hmm, lot's of [technical weasel words](https://en.wikipedia.org/wiki/Weasel_word) in that response.

As a user, I want to know specifically what happens in plain English.

So for the requesting adaptor. What response did I get back to my original facade request? Did the transfer get sent to the sending system? If I got a failed response, can I request the patient again?

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (adrianclay) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:45]: Oh strange. How could any documents get saved, if the file storage is down. This test seems very unlikely?!

Documentation which is wrong is worse than no documentation.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 2 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md:14]: ok, I will reword the subtitle based on your suggestion

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 3 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: in NIAD-2369 in the file attached for tests it says "Queue" but I guess it meant to be more like the broker is down.

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 4 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: Made changes based on your feedback

File: dependent-service-downtime-and-ps-adaptor-behavior.md (hunk 5 above)
Comment (ole4ryb) [reply dependent-service-downtime-and-ps-adaptor-behavior.md]: I will rephrase it

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
+## Service Failures and Expected Behavior
+
+This guide outlines the expected behavior of the PS Adaptor when various dependent services are down. 
+The PS Adaptor works in conjunction with several services including the GP2GP Translator, MHS Adaptor, and others.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:4]: What do you mean `GP2GP Translator` in this context?

**Comment by adrianclay** [comment]
We should be referring to the adaptor as the "Requesting adaptor throughout"

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+   - Scenario: The GP2GP Translator is not operational and cannot send patient records.
+   - Expected Behavior: No messages are sent to the incumbent system while the translator is down. Once the GP2GP Translator recovers, 
+     the transfer process resumes, and messages are processed as normal.
+   - Recovery: Automatic upon Translator recovery; transfer resumes without manual intervention.
+2. Message Broker is Down
+   - Scenario: The message broker responsible for transferring data between the GP2GP Translator and Requesting Adaptor is down.
+   - Expected Behavior: The initial request is never sent to the incumbent system because the queue is unavailable.
+                        The facade responds with 500 "Internal Server Error". The transfer wasn't sent to the sending adapter.
+                        The transfer can be requested again once the message broker is up and running.
+   - Recovery: After the queue is restored, the request can be processed normally, but any messages queued during downtime are not sent.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:25]: This only talks about the failure in the context of starting a new conversation. It doesn't cover what happens to transfers which are currently in progress.

File: dependent-service-downtime-and-ps-adaptor-behavior.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+3. Requesting Adaptor Database (DB) is Down
+   - Scenario: The Requesting Adaptor's database is not operational.
+   - Expected Behavior: The initial transfer request is not sent because the adaptor cannot access the necessary data.
+   - Recovery: After the database is restored the transfer is processed as normal.
```
Comment (adrianclay) [inline dependent-service-downtime-and-ps-adaptor-behavior.md:29]: Is this correct? I would expect the NME to have to request the transfer again.

**Review by adrianclay — APPROVED** [review]
